swarm/storage/pushsync/pushsync_simulation_test.go

```diff
@@ -0,0 +1,280 @@
+// * for each trial, two nodes are chosen randomly, an uploader and a downloader
+// * uploader uploads a number of chunks
+// * wait until the uploaded chunks are synced
+// * downloader downloads the chunk
+// Trials are run concurrently
+func TestPushSyncSimulation(t *testing.T) {
+	nodeCnt := 64
+	chunkCnt := 1000
+	trials := 10
+	err := testSyncerWithPubSub(nodeCnt, chunkCnt, trials, newServiceFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testSyncerWithPubSub(nodeCnt, chunkCnt, trials int, sf simulation.ServiceFunc) error {
+	sim := simulation.New(map[string]simulation.ServiceFunc{
+		"streamer": sf,
+	})
+	defer sim.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	err := sim.UploadSnapshot(ctx, fmt.Sprintf("../../network/stream/testing/snapshot_%d.json", nodeCnt))
+	if err != nil {
+		return err
+	}
+	log.Info("Snapshot loaded")
+
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		errc := make(chan error)
+		for i := 0; i < trials; i++ {
+			go uploadAndDownload(ctx, sim, errc, nodeCnt, chunkCnt, i)
+		}
+		i := 0
+		for err := range errc {
+			if err != nil {
+				return err
+			}
+			i++
+			if i >= trials {
+				break
+			}
+		}
+		return nil
+	})
+
+	if result.Error != nil {
+		return fmt.Errorf("simulation error: %v", result.Error)
+	}
+	log.Info("PASS")
+	return nil
+}
+
+// pickNodes selects 2 distinct
+func pickNodes(n int) (i, j int) {
+	i = rand.Intn(n)
+	j = rand.Intn(n - 1)
+	if j >= i {
+		j++
+	}
+	return
+}
+
+func uploadAndDownload(ctx context.Context, sim *simulation.Simulation, errc chan error, nodeCnt, chunkCnt, i int) {
+	// chose 2 random nodes as uploader and downloader
+	u, d := pickNodes(nodeCnt)
+	// setup uploader node
+	uid := sim.UpNodeIDs()[u]
+	val, _ := sim.NodeItem(uid, bucketKeyPushSyncer)
+	p := val.(*Pusher)
+	// setup downloader node
+	did := sim.UpNodeIDs()[d]
+	// the created tag indicates the uploader and downloader nodes
+	tagname := fmt.Sprintf("tag-%v-%v-%d", label(uid[:]), label(did[:]), i)
+	log.Debug("uploading", "peer", uid, "chunks", chunkCnt, "tagname", tagname)
+	tag, what, err := upload(ctx, p.store.(*localstore.DB), p.tags, tagname, chunkCnt)
+	if err != nil {
+		select {
+		case errc <- err:
+		case <-ctx.Done():
+			return
+		}
+		return
+	}
+	// wait till synced
+	for {
+		n, total, err := tag.Status(chunk.StateSynced)
+		if err == nil && n == total {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	log.Debug("synced", "peer", uid, "chunks", chunkCnt, "tagname", tagname)
+	log.Debug("downloading", "peer", did, "chunks", chunkCnt, "tagname", tagname)
+	val, _ = sim.NodeItem(did, bucketKeyNetStore)
+	netstore := val.(*storage.NetStore)
+	select {
+	case errc <- download(ctx, netstore, what):
+	case <-ctx.Done():
+	}
+	log.Debug("downloaded", "peer", did, "chunks", chunkCnt, "tagname", tagname)
+
+}
+
+// This  service fun construcs the special minimum service needed:
```

Comment:
You so sure it's "fun" ?

Reply:
Rewrote the godoc, I think it is fine now.